### ISSUE TITLE
Race conditions during rebar bare_compile on elixir 1.13

### ISSUE THREAD

**Issue by lukaszsamson** [elixir-lang/elixir#11530]
### Environment

* Elixir & Erlang/OTP versions (elixir --version): 
dev: Erlang/OTP 24 [erts-12.1.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit]
Elixir 1.13.1 (compiled with Erlang/OTP 24)
CI: cimg/elixir:1.13-node
* Operating system:
dev: macOS 11.6
CI: debian bullseye
* hex: 1.0.1
* rebar3: 3.15.2
* rebar 2.6.2

### Current behavior

When compiling deps for an app with `{:erlzmq, "~> 4.0", hex: :erlzmq_dnif}` the compilation often fails
```
* Getting erlzmq (Hex package)
===> Fetching pc v1.12.0
===> Analyzing applications...
===> Compiling pc
===> Analyzing applications...
===> Compiling verl
===> Compiling hex_core
===> Compiling rebar3_hex
===> Compiling c_src/erlzmq_nif.c
===> Linking /Users/lukaszsamson/bug_repro/erlz/deps/erlzmq/priv/erlzmq_nif.so
===> Analyzing applications...
===> Compiling erlzmq
===> Missing artifact priv/erlzmq_nif.so
** (Mix) Could not compile dependency :erlzmq, "/Users/lukaszsamson/.mix/rebar3 bare compile --paths /Users/lukaszsamson/bug_repro/erlz/_build/dev/lib/*/ebin" command failed. Errors may have been logged above. You can recompile this dependency with "mix deps.compile erlzmq", update it with "mix deps.update erlzmq" or clean it with "mix deps.clean erlzmq"
```
The Missing artifact priv/erlzmq_nif.so error is invalid as the so is created and the subsequent compile succeeds

The problem is non deterministic on dev machine but 100% reproducible on CI (most likely due to 1 threaded execution)

To reproduce
1. clone https://github.com/lukaszsamson/rebar_compile_repro
2. mix deps.get
3. mix deps.compile

if not reproduced until now, run

4 ./repro.sh

### Expected behavior

No race conditions as on elixir 1.12

**Comment by ericmj** [comment]
If this is due to a concurrency issue like you suggest then I doubt there is anything we can do about it from the Elixir side. Mix simply shells out to rebar and has no control over aspects like that.

**Comment by lukaszsamson** [comment]
@ericmj I tried reproducing it in rebar3 https://github.com/lukaszsamson/erlzmq with no luck

> Mix simply shells out to rebar and has no control over aspects like that.

Yet I'm not able to reproduce it on elixir 1.12.3, same OTP/hex/rebar

**Comment by josevalim** [comment]
We changed only two things on v1.13, which is to set the prod option and the directory rebar3 should output to. Check deps.compile.ex and try commenting some of those env vars.

**Comment by nifoc** [comment]
Just commenting that I think I have the same issue (with `re2`, not `erlzmq`): https://github.com/dukesoferl/re2/issues/62

Things that I have observed so far:

* Can't reproduce it when using `rebar3` directly
* Can't reproduce it with Elixir 1.12
* 100% reproducible with Elixir 1.13

I will look into the things that José suggested and report back if I find anything.

**Comment by lukaszsamson** [comment]
@josevalim the change from [this commit](https://github.com/elixir-lang/elixir/commit/f4f44e209c8ebce60e85d35427f02d4e305c1d8b#diff-da9fa810e81e1fe6ac8d4defde11dc457f0066fc9e442e85a96ea2e182414eac) seems to have broken it

I tested deps.compile.ex with
1. `{"REBAR_BARE_COMPILER_OUTPUT_DIR", dep_path},` - no issues
2. not setting REBAR_BARE_COMPILER_OUTPUT_DIR (as on (1.12 branch)[https://github.com/elixir-lang/elixir/commit/05ab231bc62cedf1d82efd4827fa9055ae0376f6#diff-da9fa810e81e1fe6ac8d4defde11dc457f0066fc9e442e85a96ea2e182414eac]) - no issue
3. `{"REBAR_BARE_COMPILER_OUTPUT_DIR", build_path},` - the issue is reproducible

I don't know the reasons behind https://github.com/elixir-lang/elixir/issues/11182 but I can tell that erlzmq writes its .so to priv dir and specifies this in rebar.config
```
{artifacts, ["priv/erlzmq_nif.so"]}.
{port_specs, [{ "priv/erlzmq_nif.so", ["c_src/*.c"] }]}.
```
and then expects to lad it from there
```
init() ->
    PrivDir = case code:priv_dir(?MODULE) of
                  {error, _} ->
                      EbinDir = filename:dirname(code:which(?MODULE)),
                      AppPath = filename:dirname(EbinDir),
                      filename:join(AppPath, "priv");
                  Path ->
                      Path
              end,
    erlang:load_nif(filename:join(PrivDir, "erlzmq_nif"), 0).
```

**Comment by josevalim** [comment]
Thank you @nifoc and @lukaszsamson! My understanding is that `REBAR_BARE_COMPILER_OUTPUT_DIR` is not being respected by the port plugin. I assume they are writing to the priv directory in the dep folder, not the one pointed out by the bare compiler. So checking the artifact fails.

**Comment by josevalim** [comment]
Here is a fix in the port compiler plugin: https://github.com/blt/port_compiler/pull/73

Thank you @starbelly! :heart:

**Comment by starbelly** [comment]
Note : 

While waiting on the patch to be merged and a release to ensue and in the case you're just trying to compile for local development purposes you can use this patched version by going into `deps/erlzmq` and editing the `rebar.config`, replacing `{plugins, [pc,rebar3_hex]}.` with `{plugins, [{pc, ".*", {git, "https://github.com/starbelly/port_compiler", {branch, "honor_rebar_bare_compiler_output_dir"}}}, rebar3_hex]}.` 

This won't help you in CI, etc. but will at least serve as a temporary work around for local dev.

**Comment by lukaszsamson** [comment]
Will the library be able to load its .so lib from REBAR_BARE_COMPILER_OUTPUT_DIR without code changes? Will it get copied/symlinked to priv?

**Comment by josevalim** [comment]
Yes, that part is already done and should work. The issue is that rebar3 was failing before we got to that.

**Comment by tsloughter** [comment]
@starbelly so this patch definitely works with erlzmq? I'm a little confused by the non-legacy code path, wonder if that isn't used by erlzmq build or if I just don't understand what it is doing :)

**Comment by starbelly** [comment]
@tsloughter  I did not test with erlzmq, just enacl, which goes down the non-legacy code path.

**Comment by nifoc** [comment]
I don't know which path `re2` goes down, but the PR fixes the issue for it, too 🥳️

**Comment by starbelly** [comment]
I

> I don't know which path `re2` goes down, but the PR fixes the issue for it, too 🥳️

Nice! I just tested that as well, it goes down the non-legacy path. 

To note for others I also tested jiffy, which goes the legacy path. I did not test erlzmq simply because I had trouble compiling (unrelated to mix, rebar3, or pc).

**Comment by starbelly** [comment]
@tsloughter merged the patch and released 1.14.0 https://hex.pm/packages/pc 

If you're using rebar3 outside the context of a mix project be sure to clear your  rebar3 hex cache (`~/.cache/rebar3/hex`) to ensure it downloads the latest version.

**Comment by lukaszsamson** [comment]
I can confirm that pc 1.14 is working for me with erlzmq.
> Be sure to clear your rebar3 hex cache (~/.cache/rebar3/hex) to ensure it downloads the latest version.

Thanks @starbelly, without it rebar was redownloading 1.12 from cache. It would save some debugging if we could warn about incompatible versions...

**Comment by starbelly** [comment]
> Thanks @starbelly, without it rebar was redownloading 1.12 from cache. It would save some debugging if we could warn about incompatible versions...

You're welcome. 

To note rebar3 will keep downloading the same version as it favors explicitness. That is, it expects the user to explicitly invoke a command to instruct rebar3 to upgrade a plugin to the latest version.  It just so happens doing so without a rebar3 project or a global rebar.config is not so easy, it's easier just to blow the cache away in some situations.

In regards to emitting a warning, it's not exactly possible really. Mix merely shells out to rebar/rebar3 to compile a dependency that depends on rebar/rebar3 as the build tool. From rebar/rebar3's and thus pc's perspective nothing was ever broken or incompatible. 

In order to emit such a warning mix would have to set an additional environment variable that rebar3 would have to check, really not even rebar but pc (i,.e, if ELIXIR_VERSION is >= 1.13.0 then warn), but that wouldn't work as there's no way to go back in time and patch old versions :) 

I'm glad everything is working for you and confident you won't have problems in Ci 🚀
